src/netdef.c

```diff
@@ -0,0 +1,307 @@
+    g_free(opts->address);
+    g_free(opts->label);
+    g_free(opts->lifetime);
+    g_free(opts);
+}
+
+static void free_route(void* ptr)
+{
+    NetplanIPRoute* route = ptr;
+    g_free(route->scope);
+    g_free(route->type);
+    g_free(route->to);
+    g_free(route->from);
+    g_free(route->via);
+    g_free(route);
+}
+
+static void free_ip_rules(void* ptr)
+{
+    NetplanIPRule* rule = ptr;
+    g_free(rule->to);
+    g_free(rule->from);
+    g_free(rule);
+}
+
+static void free_wireguard_peer(void* ptr)
+{
+    NetplanWireguardPeer* wg = ptr;
+    g_free(wg->endpoint);
+    g_free(wg->preshared_key);
+    g_free(wg->public_key);
+    g_free(wg);
+}
+
+static void reset_auth_settings(NetplanAuthenticationSettings* auth)
+{
+    FREE_AND_NULLIFY(auth->identity);
+    FREE_AND_NULLIFY(auth->anonymous_identity);
+    FREE_AND_NULLIFY(auth->password);
+    FREE_AND_NULLIFY(auth->ca_certificate);
+    FREE_AND_NULLIFY(auth->client_certificate);
+    FREE_AND_NULLIFY(auth->client_key);
+    FREE_AND_NULLIFY(auth->client_key_password);
+    FREE_AND_NULLIFY(auth->phase2_auth);  /* netplan-feature: auth-phase2 */
+    auth->key_management = NETPLAN_AUTH_KEY_MANAGEMENT_NONE;
+    auth->eap_method = NETPLAN_AUTH_EAP_NONE;
+}
+
+static void reset_ovs_settings(NetplanOVSSettings* settings)
+{
+    settings->mcast_snooping = FALSE;
+    settings->rstp = FALSE;
+
+    free_hashtable_with_destructor(&settings->external_ids, g_free);
+    free_hashtable_with_destructor(&settings->other_config, g_free);
+
+    FREE_AND_NULLIFY(settings->lacp);
+    FREE_AND_NULLIFY(settings->fail_mode);
+
+    free_garray_with_destructor(&settings->protocols, g_free);
+    reset_auth_settings(&settings->ssl);
+
+    free_garray_with_destructor(&settings->controller.addresses, g_free);
+    FREE_AND_NULLIFY(settings->controller.connection_mode);
+}
+
+static void
+reset_dhcp_overrides(NetplanDHCPOverrides* overrides)
+{
+    overrides->use_dns = TRUE;
+    FREE_AND_NULLIFY(overrides->use_domains);
+    overrides->use_ntp = TRUE;
+    overrides->send_hostname = TRUE;
+    overrides->use_hostname = TRUE;
+    overrides->use_mtu = TRUE;
+    overrides->use_routes = TRUE;
+    FREE_AND_NULLIFY(overrides->hostname);
+    overrides->metric = NETPLAN_METRIC_UNSPEC;
+}
+
+/* Reset a backend settings object. The caller needs to specify the actual backend as it is not
+ * contained within the object itself! */
+static inline void reset_backend_settings(NetplanBackendSettings* settings, NetplanBackend backend)
+{
+    switch (backend) {
+        case NETPLAN_BACKEND_NETWORKD:
+            FREE_AND_NULLIFY(settings->networkd.unit);
+            break;
+        case NETPLAN_BACKEND_NM:
+            FREE_AND_NULLIFY(settings->nm.name);
+            FREE_AND_NULLIFY(settings->nm.uuid);
+            FREE_AND_NULLIFY(settings->nm.stable_id);
+            FREE_AND_NULLIFY(settings->nm.device);
+            FREE_AND_NULLIFY(settings->nm.passthrough);
```

Comment:
haha, indeed. that `GData*` type isn't documented very well. And yes, `g_datalist_clear()` seems to be the correct tool for the job here.
But we should probably Nullify the `settings->nm.passthrough` field afterwards, to avoid a dangling pointer.